Docs

jjustison opened the pull request:
I included materials from the wiki into the "Continuous trait analysis" page.  The materials are their own section at the bottom. I tried adding some explanations of the code and how one might interpret the results. Feel free to change as needed. 

One thing of note is that the fish phylogeny is not time consistent which is important for the Pagel's lambda models. For now, I added a note that users should follow the warning and calibrate their phylogenies prior to analysis. However, I then noted that in our case it is probably fine since the inconsistency between paths is minor. Does this seem sufficient or should we calibrate the network so it is time consistent?

pbastide commented:
Thank you for the addition !

I made a pass, doing some re-wording and editing. Feel free to revert if you disagree.

I wonder if this should be a separate section, instead of being nested inside the "Continuous trait analysis" section ?
We could even have a "chapter" called "Examples", with this section, and maybe others from discrete traits as they come along ? As is done e.g. [here](https://docs.rxinfer.com/stable/examples/overview/)

For the time inconsistency : it does seem that the difference is small. I did not have time to check this in details, but maybe we should have some kind of "tolerance" argument to decide consistency ?

cecileane commented:
PhyloPlots does not require any particular version of R.
I ran into the same issue for a [PR to PhyloNetworks](https://github.com/JuliaPhylo/PhyloNetworks.jl/pull/225). I deleted the github-action cache for the check that failed, from 2 weeks ago when R's version was 4.4.2 was installed by `r-lib/actions/setup-r@v2`. I then re-ran the documentation build, and it passed. So the problem seems to be that the cache remembers about R v4.4.2, and doesn't find it.

cecileane commented:
I deleted all github-action caches under Linux, and re-ran the documentation. it worked!

cecileane commented:
Example section: I like this idea. But we should think about how to split the documentation into types of analyses (current organization: continuous / discrete, analyze / simulate) versus examples dealing with the usual aspects of real data analysis (missing data, interpretation etc.). Here, this example is specifically about continuous traits. If we start a new Example section, it could grow very large later with examples of discrete traits. In the end, we would run into the same issue we would like to avoid here: a very large section.

cecileane commented:
> For the time inconsistency: I believe the function does allow a tolerance 1e-8 but perhaps the tolerance could be an explicit argument?

Agreed, that would be best, but perhaps this needs to be worked out separately, with a change to functions in PhyloNetworks so that PhyloTraits can use a tolerance option.

cecileane commented:
The documentation shows a ton of warnings: many about time-consistence, and many of these: `Warning: These prediction intervals show uncertainty in ancestral values,...`
Would you see a way to suppress these warnings? Say, show the first one of each kind, but delete most others that come after? They are distracting.

jjustison commented:
> The documentation shows a ton of warnings: many about time-consistence, and many of these: `Warning: These prediction intervals show uncertainty in ancestral values,...` Would you see a way to suppress these warnings? Say, show the first one of each kind, but delete most others that come after? They are distracting.

For a perhaps hacky solution, I found the `Suppressor.jl` package for suppressing warnings. When using this package, we can wrap any code that produces a warning in a block for suppressing warnings. E.g., to suppress the warning at line 220 of `phyloregression.md` we would modify the block (and add `using Suppressor` in the `setup` block):
```@example tree_trait
datTrait1 = DataFrame(trait1 = dat[:,:trait1], tipnames = dat[:,:tipnames])
@suppress_err begin # hide
global ancTrait1Approx = 1 # hide
ancTrait1Approx = ancestralreconstruction(datTrait1, truenet)
nothing # hide
end # hide
```
**Note:** For any variables we specify within the `suppress_err` block, we need to denote them as global variables so that they are defined outside of the scope of this block. 

Another option would be to go into the functions themselves and add a `verbose` argument for whether they print out warnings. Any preference for how to go about this?

jjustison commented:
I actually did a combination of both things:
- for the functions `phylolm`, `fitdiscrete`, `check_matchtaxonnames!`, and `ancestralreconstruction`, I added an optional keyword `verbose` argument. Some of these functions are heavily overloaded and although I added the argument to many of the methods, will only affect those that actually produce a warning.

- for warnings outside of our direct control (e.g., the GLM warning produced when using `ftest`), I used `Suppressor` to suppress those warnings

pbastide commented:
I like the `verbose` option. However, I see two potential drawbacks / conflicts : 
- The `verbose` option was already used in `fitdiscrete`, with [default to `false`](https://github.com/JuliaPhylo/PhyloTraits.jl/blob/main/src/fit_discrete.jl#L319). It seems to be linked with [debugging](https://github.com/JuliaPhylo/PhyloTraits.jl/blob/main/src/fit_discrete.jl#L615) messages. The new `verbose` option might conflict with it ?
- The `verbose`option needs to be added in many places in the code, with might make it harder to maintain. For instance [here](https://github.com/JuliaPhylo/PhyloTraits.jl/pull/6/commits/651f5b4c13a196ac695b4697c7961770790ba11e#diff-397660738f0e21fa98f06fa48ff87dad390a28fc389cf3c016b51eeceabb8032R528), the `verbose` option of `ancestralreconstruction` might clash with the one hidden in `kwargs` and passed to `phylolm` ?

I'm not sure what we could do about this, but here are possible directions:
- hide the `verbose` argument in optional `kwargs` as is done in `fitdiscrete`
- change some of the `@warn` to `@info`
- use the base [`Logging`](https://docs.julialang.org/en/v1/stdlib/Logging/) to globally suppress some messages (e.g., @info level) ?

This is just brainstorming, I'll keep thinking about this.

jjustison commented:
Good points! Yes, this might take some more thought.

 I like the idea of hiding the suppression warning argument in `kwargs` where it is not explicitly used. This should make maintenance a little easier. 

And good catch about conflicting uses of `verbose`. For `fitIdiscrete`, it looks like `verbose` wasn't used directly in the function but passed to the `fit!` function to display more information. Maybe to avoid conflict, we could use `verbose` strictly for displaying extra info and then some other argument (for example, `suppresswarnings`)?

pbastide commented:
I tried to change `verbose` to `showWarnings`. What do you think about it ?

cecileane commented:
Thank you both!
- could we use `showwarnings` in all-small-case?
- Suppressor doesn't seem to fully work. The website [preview](https://juliaphylo.github.io/PhyloTraits.jl/previews/PR6/) has a lot of these `::warning file=../../../.julia/packages/GLM/vM20T/src/lm.jl,line=185::Starting from GLM.jl 1.8, null model is defined as having no predictor at all when a model without an intercept is passed.` around [here](https://juliaphylo.github.io/PhyloTraits.jl/previews/PR6/man/phyloregression/#Transgressive-evolution) for example.

cecileane commented:
The custom logger works very well!

cecileane commented:
Back to a separate "Examples" section: do you think it could be at the same level as "Manual" and "Library"? 
- The menu on the left is from `docs/make.jl`.
- The table of content listed on the home page is controlled by `docs/src/index.md`.
- The source `.md` files for these examples could still be in `docs/src/man`.

If we have an "Examples" section at the same level as the "Manual" section, the subsections under "Examples" could more or less follow the same organization as those in the manual, but contain more tips about how to deal with messy data, and interpretation.

I'm happy to try and go that route if you think it's a good idea. But there could be better options, including the status quo. What do you think?

cecileane commented:
So, I am afraid that we have a misunderstanding here. I don't think that we need an option to suppress warnings, always. I was only thinking of 2 types of warnings:

1. from ancestral state [reconstruction](https://github.com/JuliaPhylo/PhyloTraits.jl/blob/dab6d991e1f3a60560355c5c4a95d80edb280799/src/ancestral_continuous.jl#L312C21-L312C26): "Warning: These prediction intervals show uncertainty in ancestral values,...". This warning is lengthy, and it's easy to run this function many times. But we don't need to see the warning every time. After we read it once, say for the sword index, we get it. After we run the function for mate preference, we don't need to read it again. If a user has 10 response variables of interest and and 3 tree/networks, they might want to suppress the warning to avoid seeing it 30 times.
2. from a [time-inconsistent network](https://github.com/JuliaPhylo/PhyloTraits.jl/blob/dab6d991e1f3a60560355c5c4a95d80edb280799/src/fit_phylolm_continuous.jl#L103): "Warning: The network is not time consistent...". Here, after seeing the warning, a user may decide to continue using their network anyway, or may decide to go back and re-calibrate their network to make it time-consistent. If the user decides that the time-inconsistency is minor or can be ignored, then they might run many `phylolm` analyses. In that case, they might want to suppress the warning to avoid seeing it over and over.

But I don't think that we need an option to suppress other warnings. For example:
- warning about non-matching tip labels between the network and the trait data, from [`check_matchtaxonnames!`](https://github.com/JuliaPhylo/PhyloTraits.jl/blob/dab6d991e1f3a60560355c5c4a95d80edb280799/src/fit_discrete.jl#L924). Do you think that this function will be used repeatedly by a user, such that they will want to use an option to suppress a repeated warning? I didn't think so, because the pruned network is returned by the function.
- or the warning when tips don't match at all, [here](https://github.com/JuliaPhylo/PhyloTraits.jl/blob/dab6d991e1f3a60560355c5c4a95d80edb280799/src/ancestral_continuous.jl#L285)? When tips don't match, it's quite a serious issue. But I didn't think enough about a use-case when a user would get this warning over and over again.

pbastide commented:
> Back to a separate "Examples" section: do you think it could be at the same level as "Manual" and "Library"?
> 
>     * The menu on the left is from `docs/make.jl`.
> 
>     * The table of content listed on the home page is controlled by `docs/src/index.md`.
> 
>     * The source `.md` files for these examples could still be in `docs/src/man`.
> 
> 
> If we have an "Examples" section at the same level as the "Manual" section, the subsections under "Examples" could more or less follow the same organization as those in the manual, but contain more tips about how to deal with messy data, and interpretation.
> 
> I'm happy to try and go that route if you think it's a good idea. But there could be better options, including the status quo. What do you think?

Yes, that is what I had in mind initially ! (with "chapters" sorry I was not clear)
I like for the examples to be separated from the rest, as they might be a first entry point for data minded people.
This also kind of reproduce the previous organisation, with the "manual" on the webpage and the "wiki" separated with examples.
This is also a bit similar to what they have in beast (https://beast.community/), with generic "tutorials" (our "manual") separated from "workshop tutorials" (our "examples"), that are more data focused, and come from workshop sessions.

Status quo is find too, but it makes the example more difficult to find.

Do we have any other examples to put in this section if we create one ?
Maybe we could add the [polemonium](https://datadryad.org/dataset/doi:10.5061/dryad.9ghx3ffkc) analysis with within-species variation ?

pbastide commented:
> So, I am afraid that we have a misunderstanding here. I don't think that we need an option to suppress warnings, always. 

I don't have a strong opinion on which warnings should stay and which should go.
A (lazy) option could also be to not add this option, but teach people how to use the custom loggers if they want to, leaving the responsibility to them.
> 
>     1. from ancestral state [reconstruction](https://github.com/JuliaPhylo/PhyloTraits.jl/blob/dab6d991e1f3a60560355c5c4a95d80edb280799/src/ancestral_continuous.jl#L312C21-L312C26): "Warning: These prediction intervals show uncertainty in ancestral values,...". This warning is lengthy, and it's easy to run this function many times. But we don't need to see the warning every time. After we read it once, say for the sword index, we get it. After we run the function for mate preference, we don't need to read it again. If a user has 10 response variables of interest and and 3 tree/networks, they might want to suppress the warning to avoid seeing it 30 times.
> 
>     2. from a [time-inconsistent network](https://github.com/JuliaPhylo/PhyloTraits.jl/blob/dab6d991e1f3a60560355c5c4a95d80edb280799/src/fit_phylolm_continuous.jl#L103): "Warning: The network is not time consistent...". Here, after seeing the warning, a user may decide to continue using their network anyway, or may decide to go back and re-calibrate their network to make it time-consistent. If the user decides that the time-inconsistency is minor or can be ignored, then they might run many `phylolm` analyses. In that case, they might want to suppress the warning to avoid seeing it over and over.

I agree with these two.

> 
> But I don't think that we need an option to suppress other warnings. For example:
> 
>     * warning about non-matching tip labels between the network and the trait data, from [`check_matchtaxonnames!`](https://github.com/JuliaPhylo/PhyloTraits.jl/blob/dab6d991e1f3a60560355c5c4a95d80edb280799/src/fit_discrete.jl#L924). Do you think that this function will be used repeatedly by a user, such that they will want to use an option to suppress a repeated warning? I didn't think so, because the pruned network is returned by the function.

Not sure about this one, but it was repeated around
[here](https://juliaphylo.github.io/PhyloTraits.jl/previews/PR6/man/fitdiscreteDNA/#likelihood-of-a-fixed-network) for d1, d2, d3, fits with different models, I think ?


>     * or the warning when tips don't match at all, [here](https://github.com/JuliaPhylo/PhyloTraits.jl/blob/dab6d991e1f3a60560355c5c4a95d80edb280799/src/ancestral_continuous.jl#L285)? When tips don't match, it's quite a serious issue. But I didn't think enough about a use-case when a user would get this warning over and over again.

I'm not sure I can think of a use case for this one : when would the ind field of the object be [0] ?

jjustison commented:
>Not sure about this one, but it was repeated around
[here](https://juliaphylo.github.io/PhyloTraits.jl/previews/PR6/man/fitdiscreteDNA/#likelihood-of-a-fixed-network) for d1, d2, d3, fits with different models, I think ?

I think this might be why `check_matchtaxonnames!` could helpful to suppress warnings. While the function does return the "fixed" phylogeny, it is an internal function that gets called by `fitdiscrete` and thus might appear many times as different models are fitted

cecileane commented:
That's a very interesting option indeed! We don't need to pick the same option for all types of warnings.

1. I agree that knowing the set of taxa with missing values is super important, because it may differ depending on the variables used in the model. So perhaps keep the option to suppress warnings in `check_matchtaxonnames!`?
2. For prediction intervals, use `maxlog=1` instead? (no function-based option to suppress this warning --there's always the option to use a custom logger).
   I do *not* think it's necessary to add that this message shows only once per session. The warning message is already very long. I've come across this behavior before (not knowing about the `maxlog=1` argument), and I understood that the message appeared only once per session without being told so. No being told in advance didn't bother me as a user.

> A (lazy) option could also be to not add this option, but teach people how to use the custom loggers if they want to, leaving the responsibility to them.

Ah, teaching people to use custom loggers is not easy, in my experience. We could have a short sentence that links to julia's documentation about [logging](https://docs.julialang.org/en/v1/stdlib/Logging/) to let people know that warnings (and information) logs can be suppressed or redirected to a file, for example.

pbastide commented:
Thanks for the feedback. 
So here is what I tried:
1. use `maxlog=1` for the interval warning.
2. did not silence the warning about unnamed tips in `ancestralreconstruction`
-> This removes the `suppresswarnings` option completely from `ancestralreconstruction`.
3. kept the `suppresswarnings` for the pagel lambda in `phylolm`
4. kept the `suppresswarnings` for `check_matchtaxonnames!` in `fitdiscrete`
-> this option needs to be added in many places on the code, as seen before.
5. changed `showWarnings` to `suppresswarnings`.

**Problem**: the `maxlog=1` things works well in my interactive repl, but for some reasons it seems to be ignored by `Documenter`, so the warnings for the intervals came back all over the place in the doc.
I'm not sure how to fix this. A "quick and dirty" way would just be to use the custom logger to manually suppress these warnings in the doc. That only works if we are sure the `maxlog` is working properly. Maybe we could add a [test](https://docs.julialang.org/en/v1/stdlib/Test/#Test.TestLogger) for this.

jjustison commented:
I think that sounds like a reasonable solution, as long as things work well interactively. Tomorrow I can look to see if I can get Documenter to play nicely with the `maxlog`, else I can add the custom logger blocks as needed and write a test for suppressing the warning after it appears once

jjustison commented:
From [this](https://discourse.julialang.org/t/interaction-between-logging-jl-juliaformatter-jl-and-documenter-jl/93504) tangentially related discussion, I am under the impression that the logger with Documenter is reset after each expression so manually suppressing warnings might be our best quick and easy solution

pbastide commented:
Thanks for looking this up ! Indeed, from this [issue](https://github.com/JuliaDocs/Documenter.jl/issues/1712) it seems that `Documenter` resets the logger at each bloc, to be able to capture the output.

cecileane commented:
there's still some links that don't work: from the Examples section to the Manual section, now separated, like "as done in the tutorial" at the start of [this](https://juliaphylo.github.io/PhyloTraits.jl/previews/PR6/man/example_fishes/#Ancestral-state-prediction) section. Could someone else take a look to fix this link, and see if others are similarly not working?

cecileane commented:
~I am okay with merging this.~ @jjustison do you want to hit the button? When editing the commit message, you should mention the addition of the keyword option `suppressmessage`, which is non-breaking.

Oh... it's a new feature that should warrant bumping the minor version.
So if you agree, we should modify `Project.toml` to say `version = "1.1.0"` first.

And then I'm okay with merging this PR!
